spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,491 @@
+S_test = applier.apply(df_test)
+
+# %%
+slice_names = [sf.name for sf in sfs]
+add_slice_labels(dl_train, spam_task, S_train, slice_names)
+add_slice_labels(dl_valid, spam_task, S_valid, slice_names)
+add_slice_labels(dl_test, spam_task, S_test, slice_names)
+
+# %% [markdown]
+# Like we saw above, we'd like to visualize examples in the slice.
+
+# %%
+polarity_df = slice_dataframe(df_valid, textblob_polarity)
+polarity_df[["text", "label"]].head()
+
+# %% [markdown]
+# ### Representation learning with slices
+
+# %% [markdown]
+# To cope with scale, we will attempt to learn and combine many slice-specific representations with an attention mechanism.
+# (For details, please see our technical report — coming soon!)
+# Using the helper, `convert_to_slice_tasks`, we initialize slice-specific tasks to learn "expert task heads" for each slice, in the style of multi-task learning.
+# The original `spam_task` now contains the attention mechanism to then combine these slice experts.
+
+# %%
+from snorkel.slicing import convert_to_slice_tasks
+
+slice_tasks = convert_to_slice_tasks(spam_task, slice_names)
+slice_tasks
+
+# %%
+slice_model = MultitaskClassifier(slice_tasks)
+
+# %% [markdown]
+# We train a single model initialized with all slice tasks.
+# We note that we can monitor slice-specific performance during training — this is a powerful way to track especially critical subsets of the data.
+#
+# _Note: This model includes more parameters (corresponding to additional slices), and therefore requires more time to train — we set `num_epochs=1` for demonstration purposes._
+
+# %%
+trainer = Trainer(n_epochs=1, lr=1e-4, progress_bar=True)
+trainer.fit(slice_model, [dl_train, dl_valid])
+
+# %% [markdown]
+# At inference time, the primary task head (`spam_task`) will make all final predictions.
+# We'd like to evaluate all the slice heads on the original task head.
+# To do this, we use our `remap_labels` API, as we did earlier.
+# Note that this time, we map each `ind` head to `None` — it doesn't make sense to evaluate these labels on the base task head.
+
+# %%
+Y_dict = dl_valid.dataset.Y_dict
+eval_mapping = {label: "spam_task" for label in Y_dict.keys() if "pred" in label}
+eval_mapping.update({label: None for label in Y_dict.keys() if "ind" in label})
+
+# %% [markdown]
+# _Note: in this toy dataset, we might not see significant gains because our dataset is so small that (i) there are few examples the train split, giving little signal to learn over, and (ii) there are few examples in the test split, making our evaluation metrics very noisy.
+# For a demonstration of data slicing deployed in state-of-the-art models, please see our [SuperGLUE](https://github.com/HazyResearch/snorkel-superglue/tree/master/tutorials) tutorials._
+
+# %%
+slice_model.score([dl_valid], remap_labels=eval_mapping, as_dataframe=True)
+
+# %% [markdown]
+# You've just defined slicing functions to monitor specific slices + improved slice-specific performance.
```

Comment:
This tutorial walked through the process of authoring slices, monitoring model performance on specific slices, and improving model performance using slice information.
This programming abstraction provides...
For more technical....